packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -0,0 +1,143 @@
+import {
+  TSESTree,
+  AST_NODE_TYPES,
+} from '@typescript-eslint/experimental-utils';
+import ts, { TypeFlags } from 'typescript';
+import {
+  isTypeFlagSet,
+  unionTypeParts,
+  isFalsyType,
+  isBooleanLiteralType,
+  isLiteralType,
+} from 'tsutils';
+import {
+  createRule,
+  getParserServices,
+  getConstrainedTypeAtLocation,
+} from '../util';
+
+// Truthiness utilities
+// #region
+const isTruthyLiteral = (type: ts.Type) =>
+  isBooleanLiteralType(type, true) || (isLiteralType(type) && type.value);
+
+const isPossiblyFalsy = (type: ts.Type) =>
+  unionTypeParts(type)
+    // PossiblyFalsy flag includes literal values, so exclude ones that
+    // are definitely truthy
+    .filter(t => !isTruthyLiteral(t))
+    .some(type => isTypeFlagSet(type, ts.TypeFlags.PossiblyFalsy));
+
+const isPossiblyTruthy = (type: ts.Type) =>
+  unionTypeParts(type).some(type => !isFalsyType(type));
+// #endregion
+
+type ExpressionWithTest =
+  | TSESTree.ConditionalExpression
+  | TSESTree.DoWhileStatement
+  | TSESTree.ForStatement
+  | TSESTree.IfStatement
+  | TSESTree.WhileStatement;
+
+type Options = [
+  {
+    ignoreRhs?: boolean;
+  }
+];
+
+export default createRule<Options, 'alwaysTruthy' | 'alwaysFalsy' | 'never'>({
```

Comment:
Is the second type parameter not able to be inferred from the parameters?

Reply:
Typescript doesn't currently support partial type param inference - either they're all inferred or they all need to be specified.  (Though that might change in the near future - Microsoft/Typescript#26349)